.github/workflows/ci.yml

```diff
@@ -1,34 +1,140 @@
 name: ci
 
 on:
+  workflow_dispatch:
   push:
     branches:
       - master
+    tags:
+      - v*
   pull_request:
     branches:
       - master
 
+env:
+  CACHEKEY_MAINLINE_ALL: "mainline-all"
+  CACHEKEY_BUILDKIT_HELPER_ALL: "buildkit-helper-all"
+
 jobs:
-  test:
+  validate:
     runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        script:
+          - ./hack/lint
+          - ./hack/validate-vendor
+          - ./hack/install-and-test
     steps:
-      - uses: docker/login-action@v1 
-        if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/master' }}
-        with:
-          username: ${{ secrets.DOCKERIO_USERNAME }}
-          password: ${{ secrets.DOCKERIO_PASSWORD }}
       -
+        name: Checkout
+        uses: actions/checkout@v2
+      -
+        name: Set up Docker Buildx
         uses: docker/setup-buildx-action@v1
-        name: Setup buildx
+      -
+        name: Run
+        run: |
+          ${{ matrix.script }}
+
+  build-base:
+    runs-on: ubuntu-latest
+    outputs:
+      name_suffix: ${{ steps.prep.outputs.name_suffix }}
+      push: ${{ steps.prep.outputs.push }}
+    steps:
+      -
+        name: Prepare
+        id: prep
+        run: |
+          NAME_SUFFIX=pr
+          PUSH=false
+          if [[ $GITHUB_REF == refs/tags/v* ]]; then
+            NAME_SUFFIX=${GITHUB_REF#refs/tags/}
+            PUSH=true
+          elif [[ $GITHUB_REF == refs/heads/* ]]; then
+            NAME_SUFFIX=$(echo ${GITHUB_REF#refs/heads/} | sed -r 's#/+#-#g')
+            PUSH=true
+          fi
+          echo ::set-output name=name_suffix::${NAME_SUFFIX}
+          echo ::set-output name=push::${PUSH}
+
+  build:
+    runs-on: ubuntu-latest
+    needs: [ build-base ]
+    strategy:
+      fail-fast: false
+      matrix:
+        target:
+          - mainline-all
+          - buildkit-helper-all
+    steps:
       -
         name: Checkout
         uses: actions/checkout@v2
       -
-        name: Validate all
+        name: Cache ${{ matrix.target }}
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.binfmt-cache/${{ matrix.target }}
```

Comment:
why is the matrix component needed in the path?

Reply:
Yes we can use a single cache